src/storage/lsm.rs

```diff
@@ -144,28 +132,24 @@ impl LSM {
 
         let mut memtable = Memtable::default();
         let mut commit_log_path = None;
-        let dir_iter = std::fs::read_dir(COMMIT_LOGS_DIR_PATH)?;
-        // Assume alphabetical order.
-        for dir_entry in dir_iter {
-            let path = dir_entry?.path();
+        for path in utils::read_dir_sorted(COMMIT_LOGS_DIR_PATH)? {
             memtable.update_from_commit_log(&path)?;
             commit_log_path = Some(path);
         }
 
-        let commit_log_path = commit_log_path.unwrap_or(new_path(COMMIT_LOGS_DIR_PATH));
+        let commit_log_path =
+            commit_log_path.unwrap_or(utils::timestamped_path(COMMIT_LOGS_DIR_PATH));
         let commit_log = OpenOptions::new()
             .create(true)
             .write(true)
             .append(true)
             .open(&commit_log_path)?;
 
-        // Assume alphabetical order.
-        let sstables: Vec<SSTable> = fs::read_dir(SSTABLES_DIR_PATH)?
-            .map(|entry_result| {
-                let path = entry_result.unwrap().path();
-                SSTable::read_from_file(path).unwrap()
-            })
+        let sstables: Result<Vec<SSTable>> = utils::read_dir_sorted(SSTABLES_DIR_PATH)?
+            .into_iter()
```

Comment:
Without `into_iter()`:

```
the method `map` exists for struct `std::vec::Vec<std::path::PathBuf>`, but its trait bounds were not satisfied
```

which I also don't understand

Reply:
`iter` results in an argument of `&PathBuf` and `into_iter` results in `PathBuf`. Since `read_from_file` wants a `PathBuf`, the borrowing variant `iter` is incompatible.

Reply:
I agree -- and `into_iter` is the right method here I believe.

The "trait bounds" error arises when you try to call `map()` on an iterable itself, rather than creating an iterator out of it first. There is some sort of trait that makes it work, which I'll learn about later. The compiler message is obscure.